controllers/topology/machineset_controller.go

```diff
@@ -0,0 +1,193 @@
+// i.e. the templates would otherwise be orphaned after the MachineSet deletion completes.
+// Note: To achieve this the reconciler sets a finalizer to hook into the MachineSet deletions.
+type MachineSetReconciler struct {
+	Client client.Client
+	// APIReader is used to list MachineSets directly via the API server to avoid
+	// race conditions caused by an outdated cache.
+	APIReader        client.Reader
+	WatchFilterValue string
+}
+
+func (r *MachineSetReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineSet{}).
+		Named("topology/machineset").
+		WithOptions(options).
+		WithEventFilter(predicates.All(ctrl.LoggerFrom(ctx),
+			predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue),
+			predicates.ResourceIsTopologyOwned(ctrl.LoggerFrom(ctx)),
+		)).
+		Complete(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	return nil
+}
+
+// Reconcile deletes referenced templates during deletion of topology-owned MachineSets.
+// The templates are only deleted, if they are not used in other MachineDeployments or MachineSets which are not in deleting state,
+// i.e. the templates would otherwise be orphaned after the MachineSet deletion completes.
+// Additional context:
+// * MachineSet deletion:
+//   * MachineSets are deleted and garbage collected first (without waiting until all Machines are also deleted)
+//   * After that, deletion of Machines is automatically triggered by Kubernetes based on owner references.
+// Note: We assume templates are not reused by different MachineDeployments, which is (only) true for topology-owned
+//       MachineDeployments.
+// We don't have to set the finalizer, as it's already set during MachineSet creation
+// in the MachineSet controller.
+func (r *MachineSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the MachineSet instance.
+	ms := &clusterv1.MachineSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, ms); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get MachineSet/%s", req.NamespacedName.Name)
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, ms.Namespace, ms.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, ms) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deletion reconciliation loop.
+	if !ms.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, ms)
+	}
+
```

Comment:
they way it's now the finalizer can be bypassed if manually removed which is acceptable, right?
Have you considered to ensure their existence here?

```
if !ms.ObjectMeta.DeletionTimestamp.IsZero() {
     return r.reconcileDelete(ctx, ms)
}

Otherwise ensure finalizer
return ctrl.Result{}, nil
```

Reply:
I think it's acceptable as it's just about template deletion and if someone wants to produce a template leak they should have it :) 

I think the idea of the current solution is that we only add the finalizer once during creation, which mainly has the advantage that the finalizer immediately applies, even when a deletion is triggered instantly.